invoice/src/main/java/org/killbill/billing/invoice/InvoiceDispatcher.java

```diff
@@ -304,10 +304,11 @@ public Invoice apply(final InvoiceModelDao input) {
                     }
                 }
             } else /* Dry run use cases */ {
-
                 final NotificationQueue notificationQueue = notificationQueueService.getNotificationQueue(DefaultInvoiceService.INVOICE_SERVICE_NAME,
                                                                                                           DefaultNextBillingDateNotifier.NEXT_BILLING_DATE_NOTIFIER_QUEUE);
-                final Iterable<NotificationEventWithMetadata<NextBillingDateNotificationKey>> futureNotifications = notificationQueue.getFutureNotificationForSearchKeys(context.getAccountRecordId(), context.getTenantRecordId());
+                final Iterable<NotificationEventWithMetadata<NextBillingDateNotificationKey>> futureNotificationsIterable = notificationQueue.getFutureNotificationForSearchKeys(context.getAccountRecordId(), context.getTenantRecordId());
+                // Copy the results as retrieving the iterator will issue a query each time. This also makes sure the underlying JDBC connection is closed.
+                final List<NotificationEventWithMetadata<NextBillingDateNotificationKey>> futureNotifications = ImmutableList.<NotificationEventWithMetadata<NextBillingDateNotificationKey>>copyOf(futureNotificationsIterable);
```

Comment:
> Hum, independently of the leak, i am wondering if returning `Iterable` where the iterator relies on querying the DB is such a good thing -- hence the copy.

Actually, the copy here was needed because we use the results in two separate places in the same function (could have worked with a rewindable `Iterator` too, but it seemed simpler and we are in a case where we shouldn't have many entries).

That being said, the API is indeed quite poor. We should have returned something that explicitly implements `Closeable` (similar to what we did for the pagination APIs).

> We have really embraced the lazy `Iterable` paradigm, but in most of our use cases, this has proven more of pain and because we end up in most (all?) situations going through (some) results, was this really such a  good choice?

We actually had started by returning a `List`, but loading the wrong page on Kaui (admin queues screen) lead to a OOM in production 😬 (sorry @andrenpaes! I think it paged you very early on that day...). That's why we later on switched to the current version (see https://github.com/killbill/killbill-commons/commit/28e3548d03dd0a1f4f2a7e15f8fe2335f8919722).

> Now with respect to the leak, i suppose this is true that the making the copy will go through all results and therefore close the connection?

Yup, to make a full copy, you need to go through all records.

Reply:
@pierre Let's discuss how we could potentially improve the api. This is just so easy to get things wrong, a recipe for disaster...

Reply:
Created https://github.com/killbill/killbill-commons/issues/31 for tracking.